pulse-rs/src/context.rs

```diff
@@ -0,0 +1,398 @@
+                                                  userdata)
+        };
+        if o.is_null() {
+            return Err(ErrorCode::from_error_code(self.errno()));
+        }
+        Ok(unsafe { operation::from_raw_ptr(o) })
+    }
+
+    pub fn subscribe<CB>(&self,
+                         m: SubscriptionMask,
+                         _: CB,
+                         userdata: *mut c_void)
+                         -> Result<Operation>
+        where CB: Fn(&Context, i32, *mut c_void)
+    {
+        debug_assert_eq!(::std::mem::size_of::<CB>(), 0);
+
+        unsafe extern "C" fn wrapped<F>(c: *mut ffi::pa_context,
+                                        success: c_int,
+                                        userdata: *mut c_void)
+            where F: Fn(&Context, i32, *mut c_void)
+        {
+            let ctx = context::from_raw_ptr(c);
+            let result = ::std::mem::transmute::<_, &F>(&())(&ctx, success, userdata);
+            std::mem::forget(ctx);
+            result
+        }
+
+        let o = unsafe {
+            ffi::pa_context_subscribe(self.raw_mut(), m.into(), Some(wrapped::<CB>), userdata)
+        };
+        if o.is_null() {
+            return Err(ErrorCode::from_error_code(self.errno()));
+        }
+        Ok(unsafe { operation::from_raw_ptr(o) })
+    }
+
+    pub fn clear_subscribe_callback(&self) {
+        unsafe {
+            ffi::pa_context_set_subscribe_callback(self.raw_mut(), None, ptr::null_mut());
+        }
+    }
+
+    pub fn set_subscribe_callback<CB>(&self, _: CB, userdata: *mut c_void)
+        where CB: Fn(&Context, SubscriptionEvent, u32, *mut c_void)
+    {
+        debug_assert_eq!(::std::mem::size_of::<CB>(), 0);
+
+        unsafe extern "C" fn wrapped<F>(c: *mut ffi::pa_context,
+                                        t: ffi::pa_subscription_event_type_t,
+                                        idx: u32,
+                                        userdata: *mut c_void)
+            where F: Fn(&Context, SubscriptionEvent, u32, *mut c_void)
+        {
+            let ctx = context::from_raw_ptr(c);
+            let event = SubscriptionEvent::try_from(t)
+            .expect("pa_context_subscribe_cb_t passed invalid pa_subscription_event_type_t");
+            let result = ::std::mem::transmute::<_, &F>(&())(&ctx, event, idx, userdata);
+            ::std::mem::forget(ctx);
+            result
+        }
+
+        unsafe {
+            ffi::pa_context_set_subscribe_callback(self.raw_mut(), Some(wrapped::<CB>), userdata);
+        }
+    }
+}
+
+impl Clone for Context {
+    fn clone(&self) -> Self {
+        unsafe { Context(ffi::pa_context_ref(self.raw_mut())) }
+    }
+}
+
+impl ::std::default::Default for Context {
+    fn default() -> Self {
+        Context(::std::ptr::null_mut())
+    }
+}
+
+impl Drop for Context {
+    fn drop(&mut self) {
+        if !self.is_null() {
+            unsafe {
+                ffi::pa_context_unref(self.raw_mut());
```

Comment:
Good question. This implementation is relying on the backend/Context to drain, disconnect, and shutdown before dropping it pulse::Context.

Should we make the unref explicit so that this can't accidentally happen behind the back of the owner?